src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -48,6 +50,259 @@ let GetInitialOptimizationEnv (tcImports: TcImports, tcGlobals: TcGlobals) =
+type Phase =
+    {
+        Idx: PhaseIdx
+        Name: string
+    }
+
+    override this.ToString() = $"{this.Idx}-{this.Name}"
+
+type PhaseInfo = { Phase: Phase; Func: PhaseFunc }
+
+[<RequireQualifiedAccess>]
+module private ParallelOptimization =
+    open Optimizer
+
+    /// Identifies a work item scheduled independently of others - consists of a (file) index and OptimizationPhase.
+    /// There are (NumberOfFiles * NumberOfPhases) nodes in the whole optimization process.
+    type private Node =
+        {
+            FileIdx: int
+            Phase: PhaseIdx
+        }
+
+        override this.ToString() = $"[{this.FileIdx}-{this.Phase}]"
+
+    /// Final processing of file results to produce output needed for further compilation steps.
+    let private collectFinalResults
+        (fileResults: PhaseRes[])
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+        let finalFileResults =
+            fileResults
+            |> Array.map (fun (file, res) ->
+                let implFile =
+                    {
+                        ImplFile = file
+                        OptimizeDuringCodeGen = res.FirstLoopRes.OptDuringCodeGen
+                    }
+
+                implFile, res.FirstLoopRes.OptInfo)
+
+        let lastFileFirstLoopEnv =
+            fileResults |> Array.last |> (fun (_file, res) -> res.FirstLoopRes.OptEnv)
+
+        finalFileResults, lastFileFirstLoopEnv
+
+    let optimizeFilesInParallel
+        (env0: IncrementalOptimizationEnv)
+        (phases: PhaseInfo[])
+        (files: CheckedImplFile list)
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+
+        let files = files |> List.toArray
+
+        /// Initial state for processing the current file.
+        let initialState =
+            {
+                FirstLoopRes =
+                    {
+                        OptEnv = env0
+                        OptInfo = lazy failwith "This dummy value wrapped in a Lazy was not expected to be evaluated before being replaced."
+                        HidingInfo = SignatureHidingInfo.Empty
+                        // A no-op optimizer
+                        OptDuringCodeGen = fun _ expr -> expr
+                    }
+                OptEnvExtraLoop = env0
+                OptEnvFinalSimplify = env0
+            }
+
+        // Functions for accessing a set of node jobs and their results
+        let getTask, setTask =
+            let tasks: Task<PhaseRes>[,] = Array2D.zeroCreate files.Length phases.Length
+            let getTask (node: Node) = tasks[node.FileIdx, node.Phase]
+            let setTask (node: Node) (task: Task<PhaseRes>) = tasks[node.FileIdx, node.Phase] <- task
+            getTask, setTask
+
+        /// Asynchronously wait for dependencies of a node
+        let getNodeInputs (node: Node) =
+            task {
```

Comment:
We probably want to use backgroundTask here instead, just to be safe? In case if this code path ends up in the in-proc code, which may be executing on UI thread.

Reply:
TBH I know very little about the difference between those.
I replaced it as you suggested now.

Reply:
Thanks, `backgroundTask` just ensures that it runs on bg thread if necessary:
https://github.com/dotnet/fsharp/blob/5b74995780a0e9d24a5db025e68b31e6d910f71f/src/FSharp.Core/tasks.fs#L238-L246